### ISSUE TITLE
@showprogress for Threads.@threads

### ISSUE THREAD

**Issue by ianshmean** [timholy/ProgressMeter.jl#135]
Would it be possible? It's probably asking a lot!
Something functionally along the lines of the example at https://github.com/timholy/ProgressMeter.jl#tips-for-parallel-programming

Perhaps `@showprogress` should also handle the `wait(t)`. Something simple like this, perhaps:

```julia
x = Array{Float64}(undef, 100)
@showprogress "Completing parallel work.." Threads.@threads for i in 1:100
    x[i] = rand(Float64)
    sleep(0.1)
end
```

**Comment by Wieke** [comment]
Using a simple spinlock seems to work, it ensures that only a single thread can call next at the same time.

```julia
    p = Progress(100)
    l = Threads.SpinLock()


    Threads.@threads for n = 1:100
        sleep(0.1)
        lock(l)
        next!(p)
        unlock(l)
    end
```

**Comment by ianshmean** [comment]
Nice! I'm certain that this test has shortcomings, but it doesn't seem to slow things down in this case:
```julia
using ProgressMeter
function foo(n)
    x = Array{Array{Float64}}(undef,n)
    a = rand(1000,1000)
    Threads.@threads for i = 1:n
        x[i] = a .* rand(1)
    end
    return x
end
function fooprogress(n)
    p = Progress(n)
    l = Threads.SpinLock()
    x = Array{Array{Float64}}(undef,n)
    a = rand(1000,1000)
    Threads.@threads for i = 1:n
        x[i] = a .* rand(1)
        lock(l)
        next!(p)
        unlock(l)
    end
    finish!(p)
    return x
end

using BenchmarkTools
@btime fooprogress(100) # 105.004 ms (535 allocations: 770.61 MiB)
@btime foo(100) # 103.154 ms (340 allocations: 770.59 MiB)
```

**Comment by ianshmean** [comment]
Or just the original `sleep(0.1)` example:
```julia
using ProgressMeter
function foo(n)
    Threads.@threads for i = 1:n
        sleep(0.1)
    end
end
function fooprogress(n)
    p = Progress(n)
    l = Threads.SpinLock()
    Threads.@threads for i = 1:n
        sleep(0.1)
        lock(l)
        next!(p)
        unlock(l)
    end
    finish!(p)
end

using BenchmarkTools
@btime fooprogress(100) # 2.043 s (1765 allocations: 117.45 KiB)
@btime foo(100) # 2.045 s (741 allocations: 20.47 KiB)
```

**Comment by ianshmean** [comment]
I've had a try at incorporating this approach: https://github.com/timholy/ProgressMeter.jl/pull/139

Thanks @Wieke!

**Comment by ianshmean** [comment]
Now available with the merging of #140. Thanks again, @Wieke 

```julia
using ProgressMeter
p = Progress(10)
Threads.@threads for i in 1:10
    sleep(2*rand())
    next!(p)
end
```

**Comment by Moelf** [comment]
is it not possible to have the nice syntax of:
```
@showprogress "Completing parallel work.." Threads.@threads for i in 1:100
```
as in the original post?
